src/circuit.rs

```diff
@@ -2,61 +2,707 @@
 
 use std::mem;
 
+use group::{Curve, GroupEncoding};
 use halo2::{
     circuit::{Layouter, SimpleFloorPlanner},
-    plonk,
+    plonk::{self, Advice, Column, Expression, Fixed, Instance as InstanceColumn, Selector},
     poly::Rotation,
     transcript::{Blake2bRead, Blake2bWrite},
 };
-use pasta_curves::{pallas, vesta};
+use pasta_curves::{
+    arithmetic::{CurveAffine, FieldExt},
+    pallas, vesta,
+};
 
 use crate::{
-    note::{nullifier::Nullifier, ExtractedNoteCommitment},
-    primitives::redpallas::{SpendAuth, VerificationKey},
+    constants::{
+        load::{NullifierK, OrchardFixedBasesFull, ValueCommitV},
+        MERKLE_DEPTH_ORCHARD,
+    },
+    keys::{
+        CommitIvkRandomness, DiversifiedTransmissionKey, NullifierDerivingKey, SpendValidatingKey,
+    },
+    note::{
+        commitment::{NoteCommitTrapdoor, NoteCommitment},
+        nullifier::Nullifier,
+        ExtractedNoteCommitment,
+    },
+    primitives::{
+        poseidon::{self, ConstantLength},
+        redpallas::{SpendAuth, VerificationKey},
+    },
+    spec::NonIdentityPallasPoint,
     tree::Anchor,
-    value::ValueCommitment,
+    value::{NoteValue, ValueCommitTrapdoor, ValueCommitment},
+};
+use gadget::{
+    ecc::{
+        chip::{EccChip, EccConfig},
+        FixedPoint, FixedPointBaseField, FixedPointShort, Point,
+    },
+    poseidon::{
+        Hash as PoseidonHash, Pow5T3Chip as PoseidonChip, Pow5T3Config as PoseidonConfig,
+        StateWord, Word,
+    },
+    sinsemilla::{
+        chip::{SinsemillaChip, SinsemillaConfig, SinsemillaHashDomains},
+        commit_ivk::CommitIvkConfig,
+        merkle::{
+            chip::{MerkleChip, MerkleConfig},
+            MerklePath,
+        },
+        note_commit::NoteCommitConfig,
+    },
+    utilities::{
+        copy,
+        plonk::{PLONKChip, PLONKConfig, PLONKInstructions},
+        CellValue, UtilitiesInstructions, Var,
+    },
 };
 
+use std::convert::TryInto;
+
 pub(crate) mod gadget;
 
 /// Size of the Orchard circuit.
-const K: u32 = 11;
+// FIXME: This circuit should fit within 2^11 rows.
+const K: u32 = 12;
+
+// Absolute offsets for public inputs.
+const ANCHOR: usize = 0;
+const NF_OLD: usize = 1;
+const CV_NET_X: usize = 2;
+const CV_NET_Y: usize = 3;
```

Comment:
```suggestion
const CV_NET_X: usize = 1;
const CV_NET_Y: usize = 2;
const NF_OLD: usize = 3;
```
This change is to match the spec:
![image](https://user-images.githubusercontent.com/643204/126510833-5fccc627-1e49-4efe-83ef-957783404c86.png)